model/map/src/main/java/org/keycloak/models/map/loginFailure/AbstractUserLoginFailureEntity.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.loginFailure;
+
+import org.keycloak.models.map.common.AbstractEntity;
+
+import java.util.Objects;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public abstract class AbstractUserLoginFailureEntity<K> implements AbstractEntity<K> {
+    private K id;
+    private String realmId;
+    private String userId;
+
+    /**
+     * Flag signalizing that any of the setters has been meaningfully used.
+     */
+    protected boolean updated;
+
+    private int failedLoginNotBefore;
+    private int numFailures;
+    private long lastFailure;
```

Comment:
In case of #7842, the transition to non-primitive types is motivated by expectation of a hierarchy on realms. Would there be a benefit of having a similar hierarchical structure for login failures?

Reply:
No, I can't think of any benefit right now, please ignore my comment.